### PR TITLE
Fix Prometheus scrape configuration parsing error

### PR DESCRIPTION
**Pull request by L3n41c**
### What does this PR do?

Fix an internal agent error when using a `prometheus_scrape.checks` config coming from a `datadog.yaml` file.

### Motivation

Fix the following error:
```
{"agent":"core","time":"2023-07-10 17:47:57 UTC","level":"WARN","file":"pkg/autodiscovery/common/utils/prometheus.go","line":"54","func":"buildInstances","msg":"Error processing prometheus configuration: json: unsupported type: map[interface {}]interface {}"}
```

### Describe how to test/QA your changes

Spawn an agent with the following Helm `values.yaml` configuration:
```yaml
agents:
  useConfigMap: true
  customAgentConfig:
    prometheus_scrape:
      enabled: true
      checks:
        - autodiscovery:
            kubernetes_annotations:
              include:
                prometheus.io/scrape: "true"
            kubernetes_container_names:
              - prom
          configurations:
            - metrics:
                - prom_counter:
                    name: counter
                    type: counter
                - prom_gauge:
                    name: gauge
                    type: gauge
              ignore_metrics_by_labels:
                series:
                  - "9"
                  - "10"
```

Create a test workload to be targeted by the configured Prometheus check:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  creationTimestamp: null
  labels:
    app: prom
  name: prom
  namespace: cons-5691
spec:
  replicas: 1
  selector:
    matchLabels:
      app: prom
  template:
    metadata:
      labels:
        app: prom
      annotations:
        prometheus.io/scrape: "true"
    spec:
      containers:
      - image: ghcr.io/datadog/apps-prometheus:main
        name: prom
        args:
          - --port
          - "8080"
        ports:
        - containerPort: 8080
```

With the current release and without this fix, the agent logs this error:
```
2023-07-26 13:53:33 UTC | CORE | WARN | (pkg/autodiscovery/common/utils/prometheus.go:53 in buildInstances) | Error processing prometheus configuration: json: unsupported type: map[interface {}]interface {}
```

and we cannot find the `openmetrics` check for the test workload.

With this fix, we shouldn’t get the error anymore and the `openmetrics` check should be properly scheduled.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by AliDatadog — APPROVED** [review]
Approving but I think it deserves a release note

File: pkg/autodiscovery/common/utils/prometheus.go [hunk 1]
```diff
@@ -48,6 +48,30 @@ func buildInstances(pc *types.PrometheusCheck, annotations map[string]string, na
...
 					}
 				}
+				// The `PrometheusCheck` config may come from two sources:
+				// Either it comes from the `DD_PROMETHEUS_SCRAPE_CHECKS` environment variable.
+				//   In this case, it has been parsed by JSON decoder
+				//     In this case, dictionaries have been decoded as `map[string]interface{}`
+				//     and everything is fine
+				// Or it comes from the `prometheus_scrape.checks` property inside the `datadog.yaml` file.
+				//   In this case, it has been parsed by YAML decoder
+				//     In this case, dictionaries have been decoded as `map[interface{}interface{}`
```
Comment (clamoriniere) [inline pkg/autodiscovery/common/utils/prometheus.go:58]:
```suggestion
				//     In this case, dictionaries have been decoded as `map[interface{}]interface{}`
```

**Review by clamoriniere — APPROVED** [review]

**Review by vickenty — APPROVED** [review]

**Review by clamoriniere — APPROVED** [review]

File: releasenotes/notes/fix_prometheus_config_parsing-9abcec2e11440c04.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+fixes:
+  - |
+    Fix a ``datadog.yaml`` configuration file parsing issue.
+    When the ``datadog.yaml`` configuration file contained a complex configuration under ``prometheus.checks[*].configurations[*].metrics``, a parsing error could lead to not have the openmetrics check properly scheduled. Instead, the agent logged the following error::
```
Comment (alai97) [inline releasenotes/notes/fix_prometheus_config_parsing-9abcec2e11440c04.yaml:12]:
```suggestion
    When the ``datadog.yaml`` configuration file contained a complex configuration under ``prometheus.checks[*].configurations[*].metrics``, a parsing error could lead to an OpenMetrics check not being properly scheduled. Instead, the Agent logged the following error::
```

**Review by alai97 — APPROVED** [review]
Looks good, one suggestion!

**Pull request merged**